### PR TITLE
Testing the Spearmint Application for Accessibility

### PR DESCRIPTION
**Pull request by astaiglesia**
Screenshot from 2021-03-23 11-11-57  
# Accessibility Testing of the Spearmint Application                                            

This pull request covers the addition of accessibility testing for the Spearmint application itself. Two tests were created utilizing Spectron's testing framework to leverage webdriverIO allowing Chromium based automation of the Spearmint application. 

Node Integration allows both test scripts to access core Electron API's to run mocha employing Chai's assertion library in tandem with Chai as Promised and Google's Accessibility Developer Tools for accessibility audits of specified electron.js and package.json files. The tests as configured utilize the Application API to automate instances of Electron running Spearmint for every integration and end to end tests that may be run. Audit results render in the IDE terminal to be addressed  by the developer.

<br>

### Integration Test Results

integration accessibility audit

<br>

### End to End Test Results

e2e accessibility audit

<br>

The issues noted by the initial audits above have been addressed along with those identified by manual testing against WCAG Standards. These changes include: 

- Providing a descriptive title element enabling a user to know the topic of a page without viewing the DOM. This addresses Google's Accessibility DevTool standard 'AX_TITLE_01' and has been implemented with reference to MDN documentation's discussion on 'Accessibility concerns' for this element. 

<p> <em> "A common navigation technique for users of assistive technology is to read the page title and infer the content the page contains." </em>

newTitle

<br>

- Providing greater contrasts between font and background colors through the website as well as including a descriptive tagline. HTML structure and Sass styling have been edited for the updated user interface and experience.

<p>  <em> “WCAG (Web Content Accessibility Guidelines) ensure that content is accessible by everyone, regardless of disability or user device. To meet these standards, text and interactive elements should have a color contrast ratio of at least 4.5:1. This ensures that viewers who cannot see the full color spectrum are able to read the text.” </em>

newMenu
newTestSuite

These changes allow Spearmint to now deliver an accessibility compliant contrast ratio of 4.71:1.

<br>

Additional tests may be created by future developers to expand scope coverage.

<hr>

## Files Created and Modified

```
| package.json
	| public/index.html
	| src/__tests__/spec.e2e.js
	| src/__tests__/spec.integra.js 
	| src/assets/stylesheets/colors.scss
	| src/pages/ProjectLoader/ProjectLoader.jsx
	| src/pages/ProjectLoader/ProjectLoader.module.scss
| tsconfig.json

```

<br>
<hr>

## Known Bug

- The corrected error still elicits an error response on re-test

<br>
<hr>

## Links to Docs

[Spectron](https://www.npmjs.com/package/spectron)

[Spectron - Accessibility](https://github.com/electron-userland/spectron#accessibility-testing)

[Electron - Accessibility](https://www.electronjs.org/docs/tutorial/accessibility)

[Chrome Accessibility Developer Tools](https://github.com/GoogleChrome/accessibility-developer-tools)

[Accessibility Dev Tools - #ax_title_01](https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_title_01)

[MDN - Title Element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/title)

[Mocha Testing Framework](https://mochajs.org/)

[Chai Assertion Library](https://www.chaijs.com/api/)

[Chai Assertions for Promises](https://www.npmjs.com/package/chai-as-promised)

<hr>

### Contributing Members

Co-authored-by: Gabriel Christo 45021525+bielchristo@users.noreply.github.com

Co-authored-by: Alfred Sta. Iglesia 75825642+astaiglesia@users.noreply.github.com

### CODE REVIEW

**Review by taoantaoan — CHANGES_REQUESTED** [review]
Great work, Alfred! This looks great. You've done a great job in implementing this.

Any chance we could double check ProjectLoader.module.scss? I wonder if there might be a couple of stray comments that we could clean up.

File: src/__tests__/spec.e2e.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+const Application = require('spectron').Application;
```
Comment (annieshinn) [inline src/__tests__/spec.e2e.js:1]: Let's think about async in this file as discussed.

File: src/pages/ProjectLoader/ProjectLoader.jsx [hunk 2]
```diff
@@ -39,18 +39,20 @@ const ProjectLoader = () => {
...
+            viewBox='0 0 24 24'
+            xmlns='http://www.w3.org/2000/svg'
+            xmlnsXlink='http://www.w3.org/1999/xlink'
+            >
+            <path
+              fill='#ffffff'
+              d='M17,8C8,10 5.9,16.17 3.82,21.34L5.71,22L6.66,19.7C7.14,19.87 7.64,20 8,20C19,20 22,3 22,3C21,5 14,5.25 9,6.25C4,7.25 2,11.5 2,13.5C2,15.5 3.75,17.25 3.75,17.25C7,8 17,8 17,8Z'
+              />
+          </svg>
+        <span id={styles.purpose}>simply testing your react application</span>
```
Comment (annieshinn) [inline src/pages/ProjectLoader/ProjectLoader.jsx:53]: Why were all these stylistic changes (indentation, removing spaces) made in this file?

**Review by annieshinn — CHANGES_REQUESTED** [review]
Hey Alfred!

As we discussed at code review, let's refactor the e2e test and bring it back to the full group! I made some comments throughout for you to look at as well. Thanks for your hard work Alfred!

Annie

File: src/pages/ProjectLoader/ProjectLoader.jsx (hunk 2 above)
Comment (annieshinn) [reply src/pages/ProjectLoader/ProjectLoader.jsx:53]: If unnecessary, please revert! I think the old way seems accurate :)

File: tsconfig.json [hunk 3]
```diff
@@ -24,5 +24,5 @@
     "src",
     "src/context/__tests__",
     "__tests__/hello.test.js"
-  ]
+, "tests/spec.integra.js"  ]
```
[the remark this answers is not in the document]
Comment (astaiglesia) [reply tsconfig.json:27]: Fixed

File: src/pages/ProjectLoader/ProjectLoader.module.scss [hunk 4]
```diff
@@ -4,10 +4,12 @@
...
   width: 100%;
   height: 50vh;
-  background-color: #02c2c3;
+  background-color: #038181;
   display: flex;
+  flex-direction: column;
   justify-content: center;
-  align-items: flex-end;
+  align-items: center;
+  // border: thin solid red;
```
[the remark this answers is not in the document]
Comment (astaiglesia) [reply src/pages/ProjectLoader/ProjectLoader.module.scss]: comments have been removed!

File: src/__tests__/spec.e2e.js (hunk 1 above)
Comment (astaiglesia) [reply src/__tests__/spec.e2e.js:1]: async issue has been addressed

File: src/pages/ProjectLoader/ProjectLoader.jsx (hunk 2 above)
Comment (astaiglesia) [reply src/pages/ProjectLoader/ProjectLoader.jsx]: Changes addressed as discussed =)

**Review by annieshinn — APPROVED** [review]
Thanks for implementing all of the changes we discussed! You're a star! ⭐ I look forward to code review in the morning, especially to learn about the new and improved test files.

Thanks Alfred!
Annie

File: src/__tests__/spec.e2e.js (hunk 1 above)
Comment (annieshinn) [reply src/__tests__/spec.e2e.js:1]: Thanks! Would love to hear about how you fixed it at code review :)

**Review by annieshinn — APPROVED** [review]
Looks good. Tests on my machine seem to function as expected. Thanks Alfred!

File: src/pages/ProjectLoader/ProjectLoader.module.scss [hunk 5]
```diff
@@ -16,7 +18,21 @@
   vertical-align: bottom;
   font-family: 'comfortaa';
   color: #fff;
-  margin-bottom: 30px;
+  padding-top: 30px;
+  margin-top: 30px;
+  flex: 1;
+  // border: thin solid red;
```
Comment (taoantaoan) [inline src/pages/ProjectLoader/ProjectLoader.module.scss]: Can we remove this comment: '// border: thin solid red;' ?

File: src/pages/ProjectLoader/ProjectLoader.module.scss [hunk 6]
```diff
@@ -16,7 +18,21 @@
...
+
+#purpose {
+  font-size: 1rem;
+  text-align: center;
+  // vertical-align: center;
+  font-family: 'comfortaa';
+  color: #fff;
+  // padding-top: 30px;
+  flex: 1;
+  // border: thin solid blue;
```
Comment (taoantaoan) [inline src/pages/ProjectLoader/ProjectLoader.module.scss]: Might we be able to remove these comments?

File: src/pages/ProjectLoader/ProjectLoader.module.scss [hunk 7]
```diff
@@ -25,6 +41,7 @@
   margin-bottom: 38px;
   margin-left: 5px;
   color: #fff;
+  // border: thin solid red;
```
Comment (taoantaoan) [inline src/pages/ProjectLoader/ProjectLoader.module.scss]: Can we remove this comment?

**Review by taoantaoan — APPROVED** [review]

**Review by taoantaoan — APPROVED** [review]

**Review by sharon-zhu — APPROVED** [review]
Thanks for the updates! Filed pulled down and tests run well.

File: src/pages/ProjectLoader/ProjectLoader.module.scss (hunk 5 above)
Comment (astaiglesia) [reply src/pages/ProjectLoader/ProjectLoader.module.scss]: got it covered!

File: src/pages/ProjectLoader/ProjectLoader.module.scss (hunk 6 above)
Comment (astaiglesia) [reply src/pages/ProjectLoader/ProjectLoader.module.scss]: removed!

File: src/pages/ProjectLoader/ProjectLoader.module.scss (hunk 7 above)
Comment (astaiglesia) [reply src/pages/ProjectLoader/ProjectLoader.module.scss]: this was addressed. thanks~

**Pull request merged**